// @vkontakte/elephize/dist/src/ts2php/renderers/_primitives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tEndOfFileToken = exports.tContinueStatement = exports.tBreakStatement = exports.tUndefinedLiteral = exports.tNullLiteral = exports.tFalseKeyword = exports.tTrueKeyword = exports.tNotEqualsStrictToken = exports.tNotEqualsToken = exports.tEqualsEqualsStrictToken = exports.tEqualsEqualsToken = exports.tEqualsToken = exports.tLessThanToken = exports.tGreaterThanToken = exports.tEqualsLessThanToken = exports.tEqualsGreaterThanToken = exports.tBinaryAndEqToken = exports.tBinaryOrEqToken = exports.tPercentEqToken = exports.tSlashEqToken = exports.tAsteriskEqToken = exports.tMinusEqToken = exports.tPlusEqToken = exports.tBinaryAndToken = exports.tBinaryOrToken = exports.tLogicAndToken = exports.tLogicOrToken = exports.tPercentToken = exports.tSlashToken = exports.tAsteriskToken = exports.tMinusToken = exports.tPlusToken = void 0;
var tPlusToken = function () { return '+'; };
exports.tPlusToken = tPlusToken;
var tMinusToken = function () { return '-'; };
exports.tMinusToken = tMinusToken;
var tAsteriskToken = function () { return '*'; };
exports.tAsteriskToken = tAsteriskToken;
var tSlashToken = function () { return '/'; };
exports.tSlashToken = tSlashToken;
var tPercentToken = function () { return '%'; };
exports.tPercentToken = tPercentToken;
var tLogicOrToken = function () { return '||'; };
exports.tLogicOrToken = tLogicOrToken;
var tLogicAndToken = function () { return '&&'; };
exports.tLogicAndToken = tLogicAndToken;
var tBinaryOrToken = function () { return '|'; };
exports.tBinaryOrToken = tBinaryOrToken;
var tBinaryAndToken = function () { return '&'; };
exports.tBinaryAndToken = tBinaryAndToken;
var tPlusEqToken = function () { return '+='; };
exports.tPlusEqToken = tPlusEqToken;
var tMinusEqToken = function () { return '-='; };
exports.tMinusEqToken = tMinusEqToken;
var tAsteriskEqToken = function () { return '*='; };
exports.tAsteriskEqToken = tAsteriskEqToken;
var tSlashEqToken = function () { return '/='; };
exports.tSlashEqToken = tSlashEqToken;
var tPercentEqToken = function () { return '%='; };
exports.tPercentEqToken = tPercentEqToken;
var tBinaryOrEqToken = function () { return '|='; };
exports.tBinaryOrEqToken = tBinaryOrEqToken;
var tBinaryAndEqToken = function () { return '&='; };
exports.tBinaryAndEqToken = tBinaryAndEqToken;
var tEqualsGreaterThanToken = function () { return '>='; };
exports.tEqualsGreaterThanToken = tEqualsGreaterThanToken;
var tEqualsLessThanToken = function () { return '<='; };
exports.tEqualsLessThanToken = tEqualsLessThanToken;
var tGreaterThanToken = function () { return '>'; };
exports.tGreaterThanToken = tGreaterThanToken;
var tLessThanToken = function () { return '<'; };
exports.tLessThanToken = tLessThanToken;
var tEqualsToken = function () { return '='; };
exports.tEqualsToken = tEqualsToken;
var tEqualsEqualsToken = function () { return '=='; };
exports.tEqualsEqualsToken = tEqualsEqualsToken;
var tEqualsEqualsStrictToken = function () { return '==='; };
exports.tEqualsEqualsStrictToken = tEqualsEqualsStrictToken;
var tNotEqualsToken = function () { return '!='; };
exports.tNotEqualsToken = tNotEqualsToken;
var tNotEqualsStrictToken = function () { return '!=='; };
exports.tNotEqualsStrictToken = tNotEqualsStrictToken;
var tTrueKeyword = function () { return 'true'; };
exports.tTrueKeyword = tTrueKeyword;
var tFalseKeyword = function () { return 'false'; };
exports.tFalseKeyword = tFalseKeyword;
var tNullLiteral = function () { return 'null'; };
exports.tNullLiteral = tNullLiteral;
var tUndefinedLiteral = function () { return 'null'; };
exports.tUndefinedLiteral = tUndefinedLiteral;
var tBreakStatement = function () { return 'break;'; };
exports.tBreakStatement = tBreakStatement;
var tContinueStatement = function () { return 'continue;'; };
exports.tContinueStatement = tContinueStatement;
var tEndOfFileToken = function () { return '\n'; };
exports.tEndOfFileToken = tEndOfFileToken;
