// @vkontakte/elephize/dist/src/ts2php/components/cjsModules/reactModule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactModule = void 0;
var commonjsModule_1 = require("./commonjsModule");
var moduleRegistry_1 = require("./moduleRegistry");
var phpParser_1 = require("./phpParser");
var path = require("path");
var ReactModule = /** @class */ (function (_super) {
    __extends(ReactModule, _super);
    function ReactModule() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isDerived = true;
        _this._phpdoc = [];
        return _this;
    }
    ReactModule.getIntrinsicReturnType = function (pathToModuleImpl) {
        if (!ReactModule._intrinsicReturnType) {
            var parser = new phpParser_1.PhpParsedStruct(pathToModuleImpl);
            ReactModule._intrinsicReturnType = parser.getRetval('render') || 'mixed'; // Hardcoded function name
        }
        return ReactModule._intrinsicReturnType;
    };
    ReactModule.prototype.setArgs = function (args) {
        // Support different var name for props:
        if (args === '') {
            args = '$props'; // Fallback: we should have props at out render method to comply with base class
            this._phpdoc = ['\t * @param mixed[] $props'];
        }
        else {
            this._phpdoc = args.split(', ').map(function (v) { return "\t * @param mixed[] " + v; });
            if (this._phpdoc.length > 1) {
                // We dont support react context
                this.log.error('React context is not supported (passed as 2nd argument)', [], this.log.shortCtx(this.sourceFileName));
            }
        }
        this.args = args;
    };
    ReactModule.prototype.setBlock = function (block) {
        this.block = block;
    };
    ReactModule.prototype.isEmpty = function () {
        return false; // block is never empty
    };
    ReactModule.prototype.generateContent = function () {
        var fullyQualifiedNamespace = moduleRegistry_1.ModuleRegistry.pathToNamespace(this.targetFileName);
        var builtinIntrinsicFile = path.resolve(this.builtinsPath, 'IntrinsicElement.php');
        var renderRetval = ReactModule.getIntrinsicReturnType(builtinIntrinsicFile);
        return "<?php\n/* NOTICE: autogenerated file; Do not edit by hand */\nnamespace " + fullyQualifiedNamespace + ";\nuse " + this._namespaces.builtins + "\\RenderableComponent;\nuse " + this._namespaces.builtins + "\\Stdlib;\n\nclass " + this.className + " extends RenderableComponent {\n    /**\n     * @var " + this.className + " $_mod\n     */\n    private static $_mod;\n    public static function getInstance(): " + this.className + " {\n        if (!self::$_mod) {\n            self::$_mod = new " + this.className + "();\n        }\n        return self::$_mod;\n    }\n    \n    " + Array.from(this._hoistedContent.values()).join('\n') + "\n    \n    private function __construct() {\n        " + this._constructorStatements.join('\n') + "\n    }\n    \n    /**\n    " + this._phpdoc.join('\n') + "\n     * @param mixed[] $children\n     * @return " + (renderRetval === 'mixed' ? 'mixed' : renderRetval) + "\n     */\n    public function render(array " + this.args + ", array $children) " + this.block + "\n}\n";
    };
    ReactModule._intrinsicReturnType = '';
    return ReactModule;
}(commonjsModule_1.CommonjsModule));
exports.ReactModule = ReactModule;
