// @vkontakte/elephize/dist/src/vendor/php-parser/language/syntax/SyntaxVisitor.Generated.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// ----------------------------------------------------------------------------
// THIS IS A GENERATED FILE. DO NOT MODIFY.
// ----------------------------------------------------------------------------
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntaxVisitor = void 0;
var SyntaxVisitor = /** @class */ (function () {
    function SyntaxVisitor() {
    }
    SyntaxVisitor.prototype.defaultVisit = function (node) {
        // Does nothing.
    };
    SyntaxVisitor.prototype.visit = function (node) {
        node.accept(this);
    };
    SyntaxVisitor.prototype.visitToken = function (token) {
        // Does nothing.
    };
    SyntaxVisitor.prototype.visitTrivia = function (trivia) {
        // Does nothing.
    };
    SyntaxVisitor.prototype.visitSourceText = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitAnonymousClass = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitAnonymousFunction = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitAnonymousObjectCreation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitArgument = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitArray = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitArrayElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitArrowFunction = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitAssignmentExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitBinaryExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitBreak = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClassConstant = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClassConstantDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClassConstantElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClassDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClone = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitClosureUse = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitConditionalExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitConstant = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitConstantDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitConstantElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitContinue = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitDeclare = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitDeclareBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitDestructuringAssignment = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitDoWhile = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitEcho = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitElementAccess = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitElse = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitElseBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitElseIf = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitElseIfBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitEmptyIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitErrorControl = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitEvalIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitExitIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitExpressionGroup = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitExpressionStatement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFlexibleHeredocElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFlexibleHeredocTemplate = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFor = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitForBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitForEach = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitForEachBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFullyQualifiedName = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFunctionDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitFunctionInvocation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitGlobalDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitGoTo = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitHaltCompiler = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitHeredocTemplate = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIf = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIfBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIncompleteMember = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIncompleteNamedTraitAdapatation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIncompleteReferencedTraitAdaptation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitMemberAccess = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitMethodInvocation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitObjectCreation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitScopedInvocation = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIndirectStringVariable = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIndirectVariable = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitInstanceOf = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitInterfaceDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitIsSetIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitLabel = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitLexicalVariable = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitListDestructure = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitListDestructureElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitLiteral = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitLocalVariable = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitMethodDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitMethodReference = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTraitAlias = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitType = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitNamespaceDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitNamespaceGroupDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitParameter = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitPartiallyQualifiedName = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitPostfixUnaryExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitPrintIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitPropertyDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitPropertyElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitRelativeName = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitReturn = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitScriptInclusion = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitShellCommandTemplate = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStatementBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStaticDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStaticElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStaticProperty = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStringElementAccess = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStringExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStringTemplate = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitStringVariable = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitSwitch = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitSwitchBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitSwitchCase = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitThrow = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTraitDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTraitPrecedence = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTraitUse = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTraitUseGroup = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTry = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTryCatch = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitTryFinally = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitUnaryExpression = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitUnset = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitUseDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitUseElement = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitUseGroupDeclaration = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitWhile = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitWhileBlock = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitYield = function (node) {
        this.defaultVisit(node);
    };
    SyntaxVisitor.prototype.visitYieldFrom = function (node) {
        this.defaultVisit(node);
    };
    return SyntaxVisitor;
}());
exports.SyntaxVisitor = SyntaxVisitor;
