// @vkontakte/elephize/dist/src/ts2php/components/cli/configureLogging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configureLogging = void 0;
var log_1 = require("../../utils/log");
function configureLogging(options) {
    var opts = {
        noOutput: false,
        baseDir: options.baseDir,
        outDir: options.outDir,
        verbosity: log_1.LogVerbosity.ERROR | log_1.LogVerbosity.WARN | log_1.LogVerbosity.WITH_CONTEXT,
        forceStderr: false,
    };
    if (options.quiet) {
        opts.verbosity = 0;
    }
    else {
        if (options.verbose) {
            opts.verbosity = opts.verbosity | log_1.LogVerbosity.INFO;
        }
        if (options.verboseUsage) {
            opts.verbosity = opts.verbosity | log_1.LogVerbosity.WITH_USAGE_GRAPH_DUMP | log_1.LogVerbosity.WITH_ELIMINATION_HINTS;
        }
        if (options.verboseTypehints) {
            opts.verbosity = opts.verbosity | log_1.LogVerbosity.WITH_TYPEHINTS;
        }
    }
    return new log_1.Logger(opts, options.printer);
}
exports.configureLogging = configureLogging;
