// @vkontakte/elephize/dist/src/vendor/php-parser/parser/PhpLexerState.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhpLexerState = void 0;
/**
 * Lexing states used when tokenizing PHP files.
 */
var PhpLexerState;
(function (PhpLexerState) {
    /**
     * The lexing state used when the lexer is within the host language that
     * contains a PHP script.
     *
     * This is the default state.
     */
    PhpLexerState[PhpLexerState["InHostLanguage"] = 0] = "InHostLanguage";
    /**
     * The default lexing state when rescanning a `StringTemplate`.
     */
    PhpLexerState[PhpLexerState["InDoubleQuote"] = 1] = "InDoubleQuote";
    /**
     * The default lexing state when rescanning a `FlexdocTemplate`.
     */
    PhpLexerState[PhpLexerState["InFlexibleHeredoc"] = 2] = "InFlexibleHeredoc";
    /**
     * The default lexing state when rescanning a `FlexdocTemplate` that starts
     * with a nowdoc label.
     */
    PhpLexerState[PhpLexerState["InFlexibleNowdoc"] = 3] = "InFlexibleNowdoc";
    /**
     * The default lexing state when rescanning a `HeredocTemplate`.
     */
    PhpLexerState[PhpLexerState["InHeredoc"] = 4] = "InHeredoc";
    /**
     * The default lexing state when rescanning a `HeredocTemplate` that starts
     * with a nowdoc label.
     */
    PhpLexerState[PhpLexerState["InNowdoc"] = 5] = "InNowdoc";
    /**
     * The lexing state used when the lexer is within a PHP script.
     *
     * @see PhpLexerState.InHostLanguage
     */
    PhpLexerState[PhpLexerState["InScript"] = 6] = "InScript";
    /**
     * The default lexing state when rescanning a `ShellCommandTemplate`.
     */
    PhpLexerState[PhpLexerState["InShellCommand"] = 7] = "InShellCommand";
    /**
     * A lexing state used when scanning an element access expression within a
     * string interpolation.
     */
    PhpLexerState[PhpLexerState["InVariableOffset"] = 8] = "InVariableOffset";
    /**
     * A lexing state used when rescanning the whitespace after a line break.
     */
    PhpLexerState[PhpLexerState["LookingForHeredocIndent"] = 9] = "LookingForHeredocIndent";
    /**
     * A lexing state used when rescanning heredoc start or end labels.
     */
    PhpLexerState[PhpLexerState["LookingForHeredocLabel"] = 10] = "LookingForHeredocLabel";
    /**
     * A lexing state used when scanning an object access expression within a
     * string interpolation.
     */
    PhpLexerState[PhpLexerState["LookingForProperty"] = 11] = "LookingForProperty";
    /**
     * A lexing state used when scanning an indirect variable name.
     */
    PhpLexerState[PhpLexerState["LookingForVariableName"] = 12] = "LookingForVariableName";
})(PhpLexerState = exports.PhpLexerState || (exports.PhpLexerState = {}));
