// @vkontakte/elephize/dist/src/vendor/php-parser/parser/PhpVersion.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhpVersionInfo = exports.PhpVersion = void 0;
/**
 * Supported versions of PHP that may be parsed.
 */
var PhpVersion;
(function (PhpVersion) {
    PhpVersion[PhpVersion["Any"] = 0] = "Any";
    PhpVersion[PhpVersion["PHP7_0"] = 1] = "PHP7_0";
    PhpVersion[PhpVersion["PHP7_1"] = 2] = "PHP7_1";
    PhpVersion[PhpVersion["PHP7_2"] = 3] = "PHP7_2";
    PhpVersion[PhpVersion["PHP7_3"] = 4] = "PHP7_3";
    PhpVersion[PhpVersion["PHP7_4"] = 5] = "PHP7_4";
    PhpVersion[PhpVersion["Latest"] = 5] = "Latest";
})(PhpVersion = exports.PhpVersion || (exports.PhpVersion = {}));
/**
 * Provides information about a PHP version.
 */
var PhpVersionInfo = /** @class */ (function () {
    function PhpVersionInfo() {
    }
    /**
     * Gets a string representation of the specified PHP version.
     */
    PhpVersionInfo.getText = function (version) {
        switch (version) {
            case PhpVersion.PHP7_0:
                return '7.0';
            case PhpVersion.PHP7_1:
                return '7.1';
            case PhpVersion.PHP7_2:
                return '7.2';
            case PhpVersion.PHP7_3:
                return '7.3';
            case PhpVersion.PHP7_4:
                return '7.4';
            default:
                return '';
        }
    };
    return PhpVersionInfo;
}());
exports.PhpVersionInfo = PhpVersionInfo;
