// @vkontakte/elephize/dist/src/vendor/php-parser/language/node/NodeTransform.Generated.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// ----------------------------------------------------------------------------
// THIS IS A GENERATED FILE. DO NOT MODIFY.
// ----------------------------------------------------------------------------
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeTransform = void 0;
var NodeTransform = /** @class */ (function () {
    function NodeTransform(defaultValue) {
        this.defaultValue = defaultValue;
    }
    NodeTransform.prototype.defaultVisit = function (node) {
        return this.defaultValue;
    };
    NodeTransform.prototype.visit = function (node) {
        return node.acceptResult(this);
    };
    NodeTransform.prototype.visitToken = function (token) {
        return this.defaultValue;
    };
    NodeTransform.prototype.visitTrivia = function (token) {
        return this.defaultValue;
    };
    NodeTransform.prototype.visitSourceText = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitAnonymousClass = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitAnonymousFunction = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitAnonymousObjectCreation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitArgument = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitArray = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitArrayElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitArrowFunction = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitAssignmentExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitBinaryExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitBreak = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClassConstant = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClassConstantDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClassConstantElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClassDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClone = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitClosureUse = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitConditionalExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitConstant = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitConstantDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitConstantElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitContinue = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitDeclare = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitDeclareBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitDestructuringAssignment = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitDoWhile = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitEcho = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitElementAccess = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitElse = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitElseBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitElseIf = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitElseIfBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitEmptyIntrinsic = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitErrorControl = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitEvalIntrinsic = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitExitIntrinsic = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitExpressionGroup = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitExpressionStatement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFlexibleHeredocElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFlexibleHeredocTemplate = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFor = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitForBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitForEach = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitForEachBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFullyQualifiedName = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFunctionDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitFunctionInvocation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitGlobalDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitGoTo = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitHaltCompiler = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitHeredocTemplate = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIf = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIfBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIncompleteMember = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIncompleteNamedTraitAdapatation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIncompleteReferencedTraitAdaptation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitMemberAccess = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitMethodInvocation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitObjectCreation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitScopedInvocation = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIndirectStringVariable = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIndirectVariable = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitInstanceOf = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitInterfaceDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitIsSetIntrinsic = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitLabel = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitLexicalVariable = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitListDestructure = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitListDestructureElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitLiteral = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitLocalVariable = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitMethodDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitMethodReference = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTraitAlias = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitType = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitNamespaceDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitNamespaceGroupDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitParameter = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitPartiallyQualifiedName = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitPostfixUnaryExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitPrintIntrinsic = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitPropertyDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitPropertyElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitRelativeName = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitReturn = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitScriptInclusion = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitShellCommandTemplate = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStatementBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStaticDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStaticElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStaticProperty = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStringElementAccess = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStringExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStringTemplate = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitStringVariable = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitSwitch = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitSwitchBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitSwitchCase = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitThrow = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTraitDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTraitPrecedence = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTraitUse = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTraitUseGroup = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTry = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTryCatch = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitTryFinally = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitUnaryExpression = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitUnset = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitUseDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitUseElement = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitUseGroupDeclaration = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitWhile = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitWhileBlock = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitYield = function (node) {
        return this.defaultVisit(node);
    };
    NodeTransform.prototype.visitYieldFrom = function (node) {
        return this.defaultVisit(node);
    };
    return NodeTransform;
}());
exports.NodeTransform = NodeTransform;
