// @vkontakte/elephize/dist/src/vendor/php-parser/language/node/NodeVisitor.Generated.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// ----------------------------------------------------------------------------
// THIS IS A GENERATED FILE. DO NOT MODIFY.
// ----------------------------------------------------------------------------
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeVisitor = void 0;
var NodeVisitor = /** @class */ (function () {
    function NodeVisitor() {
    }
    NodeVisitor.prototype.defaultVisit = function (node) {
        // Does nothing.
    };
    NodeVisitor.prototype.visit = function (node) {
        node.accept(this);
    };
    NodeVisitor.prototype.visitToken = function (token) {
        // this.defaultVisit(token);
    };
    NodeVisitor.prototype.visitTrivia = function (trivia) {
        // this.defaultVisit(trivia);
    };
    NodeVisitor.prototype.visitSourceText = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitAnonymousClass = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitAnonymousFunction = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitAnonymousObjectCreation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitArgument = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitArray = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitArrayElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitArrowFunction = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitAssignmentExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitBinaryExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitBreak = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClassConstant = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClassConstantDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClassConstantElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClassDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClone = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitClosureUse = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitConditionalExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitConstant = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitConstantDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitConstantElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitContinue = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitDeclare = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitDeclareBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitDestructuringAssignment = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitDoWhile = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitEcho = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitElementAccess = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitElse = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitElseBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitElseIf = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitElseIfBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitEmptyIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitErrorControl = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitEvalIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitExitIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitExpressionGroup = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitExpressionStatement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFlexibleHeredocElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFlexibleHeredocTemplate = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFor = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitForBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitForEach = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitForEachBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFullyQualifiedName = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFunctionDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitFunctionInvocation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitGlobalDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitGoTo = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitHaltCompiler = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitHeredocTemplate = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIf = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIfBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIncompleteMember = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIncompleteNamedTraitAdapatation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIncompleteReferencedTraitAdaptation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitMemberAccess = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitMethodInvocation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitObjectCreation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitScopedInvocation = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIndirectStringVariable = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIndirectVariable = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitInstanceOf = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitInterfaceDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitIsSetIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitLabel = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitLexicalVariable = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitListDestructure = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitListDestructureElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitLiteral = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitLocalVariable = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitMethodDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitMethodReference = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTraitAlias = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitType = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitNamespaceDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitNamespaceGroupDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitParameter = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitPartiallyQualifiedName = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitPostfixUnaryExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitPrintIntrinsic = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitPropertyDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitPropertyElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitRelativeName = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitReturn = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitScriptInclusion = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitShellCommandTemplate = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStatementBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStaticDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStaticElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStaticProperty = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStringElementAccess = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStringExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStringTemplate = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitStringVariable = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitSwitch = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitSwitchBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitSwitchCase = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitThrow = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTraitDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTraitPrecedence = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTraitUse = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTraitUseGroup = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTry = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTryCatch = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitTryFinally = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitUnaryExpression = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitUnset = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitUseDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitUseElement = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitUseGroupDeclaration = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitWhile = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitWhileBlock = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitYield = function (node) {
        this.defaultVisit(node);
    };
    NodeVisitor.prototype.visitYieldFrom = function (node) {
        this.defaultVisit(node);
    };
    return NodeVisitor;
}());
exports.NodeVisitor = NodeVisitor;
