// @vkontakte/elephize/dist/src/vendor/php-parser/language/node/TriviaNode.js
/**
 * Copyright 2017 Matt Acosta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkippedTokenNode = exports.TriviaNode = void 0;
var php_common_1 = require("../../../php-common");
var Node_1 = require("./Node");
/**
 * A token that is not part of the language syntax.
 */
var TriviaNode = /** @class */ (function (_super) {
    __extends(TriviaNode, _super);
    /**
     * Constructs a `TriviaNode` object.
     *
     * @param {TokenKind} kind
     *   The type of the trivia token.
     * @param {number} fullWidth
     *   The width of the trivia token.
     * @param {ReadonlyArray<SyntaxDiagnostic>=} diagnostics
     *   A list of diagnostics associated with the trivia token.
     */
    function TriviaNode(kind, fullWidth, diagnostics) {
        var _this = _super.call(this, diagnostics || Node_1.Node.EmptyDiagnosticList) || this;
        _this._flags = 0 /* None */;
        _this._fullWidth = fullWidth;
        _this.hash = 0;
        _this.kind = kind;
        if (diagnostics !== void 0 && diagnostics.length > 0) {
            _this._flags = _this._flags | 1 /* ContainsDiagnostics */;
        }
        return _this;
    }
    Object.defineProperty(TriviaNode.prototype, "flags", {
        /**
         * @inheritDoc
         */
        get: function () {
            return this._flags;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TriviaNode.prototype, "fullWidth", {
        /**
         * @inheritDoc
         */
        get: function () {
            return this._fullWidth;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TriviaNode.prototype, "isTrivia", {
        /**
         * @inheritDoc
         */
        get: function () {
            return true;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TriviaNode.prototype, "leadingTriviaWidth", {
        /**
         * @inheritDoc
         */
        get: function () {
            return 0;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.accept = function (visitor) {
        visitor.visitTrivia(this);
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.acceptResult = function (visitor) {
        return visitor.visitTrivia(this);
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.childAt = function (index) {
        throw new php_common_1.InvalidOperationException('Unreachable');
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.createSyntaxNode = function (parent, offset) {
        throw new php_common_1.InvalidOperationException('Unreachable');
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.equals = function (value) {
        // IMPORTANT: This is a performance critical method.
        if (this === value) {
            return true;
        }
        if (this.kind === value.kind && this._fullWidth === value.fullWidth && this._flags === value.flags) {
            return true;
        }
        return false;
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.hashCode = function () {
        // IMPORTANT: This is a performance critical method.
        if (this.hash === 0) {
            this.hash = TriviaNode.prototype.computeHashCode.call(this);
        }
        return this.hash;
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.withDiagnostics = function (diagnostics) {
        return new TriviaNode(this.kind, this._fullWidth, diagnostics);
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.computeHashCode = function () {
        return php_common_1.Hash.combine(this._fullWidth, this._flags ^ (this.kind + 8192));
    };
    /**
     * @inheritDoc
     */
    TriviaNode.prototype.updateFlagsAndWidth = function (flags, fullWidth) {
        throw new php_common_1.InvalidOperationException('Unreachable');
    };
    return TriviaNode;
}(Node_1.Node));
exports.TriviaNode = TriviaNode;
/**
 * A token that was skipped by the parser.
 *
 * Skipped token nodes occur when neither the current parse context, nor any
 * enclosing contexts know how to handle a token.
 */
var SkippedTokenNode = /** @class */ (function (_super) {
    __extends(SkippedTokenNode, _super);
    /**
     * Constructs a `SkippedTokenNode` object.
     *
     * @param {TokenKind} kind
     *   The type of the skipped token.
     * @param {number} fullWidth
     *   The width of the skipped token.
     * @param {ReadonlyArray<SyntaxDiagnostic>=} diagnostics
     *   A list of diagnostics associated with the skipped token.
     */
    function SkippedTokenNode(kind, fullWidth, diagnostics) {
        var _this = _super.call(this, kind, fullWidth, diagnostics) || this;
        _this._flags = _this._flags | 2 /* ContainsSkippedText */;
        return _this;
    }
    /**
     * @inheritDoc
     */
    SkippedTokenNode.prototype.withDiagnostics = function (diagnostics) {
        return new SkippedTokenNode(this.kind, this._fullWidth, diagnostics);
    };
    return SkippedTokenNode;
}(TriviaNode));
exports.SkippedTokenNode = SkippedTokenNode;
